notebooks/02_model/hypertune_spark_deep_dive.ipynb

```diff
@@ -0,0 +1,983 @@
+    "The objective function shown below demonstrates a RMSE loss for an ALS recommender. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 14,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Customize an objective function\n",
+    "def objective(params):\n",
+    "    rank = params['rank']\n",
+    "    reg = params['reg']\n",
+    "    train = params['train'] \n",
+    "    valid = params['valid'] \n",
+    "    col_user = params['col_user'] \n",
+    "    col_item = params['col_item']\n",
+    "    col_rating = params['col_rating'] \n",
+    "    col_prediction = params['col_prediction'] \n",
+    "    k = params['k']\n",
+    "    relevancy_method = params['relevancy_method']\n",
+    "    \n",
+    "    als = ALS(\n",
+    "        rank=rank,\n",
+    "        maxIter=15,\n",
+    "        implicitPrefs=False,\n",
+    "        alpha=0.1,\n",
+    "        regParam=reg,\n",
+    "        coldStartStrategy='drop',\n",
+    "        nonnegative=False,\n",
+    "        seed=0,\n",
+    "        **HEADER_ALS\n",
+    "    )\n",
+    "    \n",
+    "    model = als.fit(train)\n",
+    "    \n",
+    "    prediction = model.transform(valid)\n",
+    "\n",
+    "    rating_eval = SparkRatingEvaluation(\n",
+    "        valid, \n",
+    "        prediction, \n",
+    "        **HEADER\n",
+    "    )\n",
+    "    \n",
+    "    rmse = rating_eval.rmse()\n",
+    "    \n",
+    "    # Return the objective function result.\n",
+    "    return {\n",
+    "       'loss': rmse,\n",
+    "       'status': STATUS_OK,\n",
+    "       'eval_time': time.time()\n",
+    "    }"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "A search space is usually defined for hyperparameter exploration. Design of search space is empirical, and depends on the understanding of how distribution of parameter of interest affects the model performance measured by the loss function. \n",
+    "\n",
+    "In the ALS algorithm, the two hyper parameters, rank and reg, affect model performance in a way that\n",
+    "* The higher the rank, the better the model performance but also the higher risk of overfitting.\n",
+    "* The reg parameter is subject to rank and it prevents overfitting in certain way. \n",
+    "\n",
+    "Therefore, in this case, a uniform distribution and a lognormal distribution sampling spaces are used for rank and reg, respectively. A narrow search space is used for illustration purpose, that is, the range of rank is from 10 to 20, while that of reg is from $e^{-5}$ to $e^{-1}$. Together with the randomly sampled hyper parameters, other parameters use for building / evaluating the recommender, like `k`, column names, data, etc., are kept as constants."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 15,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# define a search space\n",
+    "space = {\n",
+    "    'rank': hp.quniform('rank', 10, 20, 5),\n",
```

Comment:
In this one we are checking 10, 15, 20. In the previous section, are we also checking this or only 10 and 20?